services/containerregistry/containerregistry.go

```diff
@@ -0,0 +1,48 @@
+package containerregistry
+
+import (
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/manicminer/hamilton/auth"
+)
+
+// ContainerRegistryClient handles communication with Azure Container Registry
+type ContainerRegistryClient struct {
+	authorizer auth.Authorizer
+	httpClient *http.Client
+	serverURL  string
+	tenantID   string
+}
+
+// NewContainerRegistryClient returns a ContainerRegistryClient
+func NewContainerRegistryClient(authorizer auth.Authorizer, serverURL string, tenantID string) *ContainerRegistryClient {
+	httpClient := http.DefaultClient
+	return &ContainerRegistryClient{
+		authorizer,
+		httpClient,
+		serverURL,
+		tenantID,
+	}
+}
+
+// WithHttpClient replaces what http client is used for communication to Azure Container Registry
+func (c *ContainerRegistryClient) WithHttpClient(httpClient *http.Client) {
+	c.httpClient = httpClient
+}
+
+func parseService(serverURL string) (*url.URL, error) {
+	scheme := "https://"
+	if strings.HasPrefix(serverURL, "https://") {
+		scheme = ""
+	}
+
+	serviceURL, err := url.Parse(fmt.Sprintf("%s%s", scheme, serverURL))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse server URL - %w", err)
+	}
+
+	return serviceURL, nil
```

Comment:
This might be better as?

```suggestion
	serviceURL, err := url.Parse(serverURL)
	if err != nil {
		return nil, fmt.Errorf("failed to parse server URL - %w", err)
	}
	if serviceURL.Scheme == "" {
		serviceURL.Scheme = "https"
	}

	return serviceURL, nil
```

Reply:
For sure! Will fix.

Reply:
@manicminer doing it your way will actually cause a `serverURL` with the string `foo.bar` to have an empty `Host` and a Path with `foo.bar`.

Reply:
Ah I see, that makes sense. Interestingly printing such a `url.URL` still yields a valid URL string :/
https://go.dev/play/p/gkaYgWvOvkm

Happy with your approach!